File: .tmp_sc4013/java_22/3949_subtree_inversion_sum_ii.java
// CONFIG class=Solution method=maxSubtreeInversionSum types=None
// LeetCode 3949 - Subtree Inversion Sum II
// https://leetcode.com/problems/subtree-inversion-sum-ii/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public long maxSubtreeInversionSum(int[][] edges, int[] nums, int k) {
        int n = nums.length;
        List<Integer>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        for (int[] edge : edges) {
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
        }
        int[] parent = new int[n];
        Arrays.fill(parent, -2);
        parent[0] = -1;
        List<Integer> order = new ArrayList<>();
        order.add(0);
        for (int i = 0; i < order.size(); i++) {
            int u = order.get(i);
            for (int v : graph[u]) {
                if (parent[v] == -2) {
                    parent[v] = u;
                    order.add(v);
                }
            }
        }
        final long infinity = 1L << 60;
        long[][] maximum = new long[n][];
        long[][] minimum = new long[n][];
        for (int oi = n - 1; oi >= 0; oi--) {
            int u = order.get(oi);
            long[] currentMax = new long[k + 1];
            long[] currentMin = new long[k + 1];
            Arrays.fill(currentMax, -infinity);
            Arrays.fill(currentMin, infinity);
            currentMax[k] = currentMin[k] = nums[u];
            for (int v : graph[u]) {
                if (parent[v] != u) continue;
                long[] nextMax = new long[k + 1];
                long[] nextMin = new long[k + 1];
                Arrays.fill(nextMax, -infinity);
                Arrays.fill(nextMin, infinity);
                for (int first = 0; first <= k; first++) {
                    if (currentMax[first] == -infinity) continue;
                    for (int childDistance = 0; childDistance <= k; childDistance++) {
                        if (maximum[v][childDistance] == -infinity) continue;
                        int second = childDistance + 1;
                        if (second > k) second = k;
                        if (first < k && second < k && first + second < k) continue;
                        int distance = Math.min(first, second);
                        long maxValue = currentMax[first] + maximum[v][childDistance];
                        long minValue = currentMin[first] + minimum[v][childDistance];
                        nextMax[distance] = Math.max(nextMax[distance], maxValue);
                        nextMin[distance] = Math.min(nextMin[distance], minValue);
                    }
                }
                currentMax = nextMax;
                currentMin = nextMin;
            }
            if (-currentMin[k] > currentMax[0]) currentMax[0] = -currentMin[k];
            if (-currentMax[k] < currentMin[0]) currentMin[0] = -currentMax[k];
            maximum[u] = currentMax;
            minimum[u] = currentMin;
        }
        long answer = -(1L << 60);
        for (long value : maximum[0]) answer = Math.max(answer, value);
        return answer;
    }
}

File: .tmp_rs4013/rust_check22/3949_subtree_inversion_sum_ii.rs
struct Solution;
// LeetCode 3949 - Subtree Inversion Sum II
// https://leetcode.com/problems/subtree-inversion-sum-ii/

impl Solution {
    pub fn max_subtree_inversion_sum(edges: Vec<Vec<i32>>, nums: Vec<i32>, k: i32) -> i64 {
        let n = nums.len();
        let k = k as usize;
        let mut graph = vec![Vec::new(); n];
        for edge in &edges {
            graph[edge[0] as usize].push(edge[1] as usize);
            graph[edge[1] as usize].push(edge[0] as usize);
        }
        let mut parent = vec![-2i32; n];
        parent[0] = -1;
        let mut order = vec![0usize];
        let mut i = 0;
        while i < order.len() {
            let u = order[i];
            for &v in &graph[u] {
                if parent[v] == -2 {
                    parent[v] = u as i32;
                    order.push(v);
                }
            }
            i += 1;
        }
        const INFINITY: i64 = 1 << 60;
        let mut maximum = vec![Vec::new(); n];
        let mut minimum = vec![Vec::new(); n];
        for &u in order.iter().rev() {
            let mut current_max = vec![-INFINITY; k + 1];
            let mut current_min = vec![INFINITY; k + 1];
            current_max[k] = nums[u] as i64;
            current_min[k] = nums[u] as i64;
            for &v in &graph[u] {
                if parent[v] != u as i32 {
                    continue;
                }
                let mut next_max = vec![-INFINITY; k + 1];
                let mut next_min = vec![INFINITY; k + 1];
                for first in 0..=k {
                    if current_max[first] == -INFINITY {
                        continue;
                    }
                    for child_distance in 0..=k {
                        if maximum[v][child_distance] == -INFINITY {
                            continue;
                        }
                        let mut second = child_distance + 1;
                        if second > k {
                            second = k;
                        }
                        if first < k && second < k && first + second < k {
                            continue;
                        }
                        let distance = first.min(second);
                        let max_value = current_max[first] + maximum[v][child_distance];
                        let min_value = current_min[first] + minimum[v][child_distance];
                        next_max[distance] = next_max[distance].max(max_value);
                        next_min[distance] = next_min[distance].min(min_value);
                    }
                }
                current_max = next_max;
                current_min = next_min;
            }
            if -current_min[k] > current_max[0] {
                current_max[0] = -current_min[k];
            }
            if -current_max[k] < current_min[0] {
                current_min[0] = -current_max[k];
            }
            maximum[u] = current_max;
            minimum[u] = current_min;
        }
        let mut answer = -(1i64 << 60);
        for &value in &maximum[0] {
            answer = answer.max(value);
        }
        answer
    }
}

fn main() {}
